gateway/webhook.go

```diff
@@ -0,0 +1,267 @@
+	}
+	return
+}
+
+func (webhook *webhookHandleT) failRequest(w http.ResponseWriter, r *http.Request, reason string, stat string) {
+	var writeKeyFailStats = make(map[string]int)
+	misc.IncrementMapByKey(writeKeyFailStats, stat, 1)
+	webhook.gwHandle.updateWriteKeyStats(writeKeyFailStats, "gateway.write_key_failed_requests")
+	logger.Debug(reason)
+	http.Error(w, reason, 400)
+	atomic.AddUint64(&webhook.gwHandle.ackCount, 1)
+}
+
+func (webhook *webhookHandleT) batchHandler(w http.ResponseWriter, r *http.Request) {
+	logger.LogRequest(r)
+	atomic.AddUint64(&webhook.gwHandle.recvCount, 1)
+
+	writeKey, ok := parseWriteKey(r)
+	if !ok {
+		webhook.failRequest(w, r, getStatus(InvalidWriteKey), "noWriteKey")
+		return
+	}
+
+	configSubscriberLock.RLock()
+	sourceDefName, ok := enabledWriteKeySourceDefMap[writeKey]
+	configSubscriberLock.RUnlock()
+	if !ok {
+		webhook.failRequest(w, r, getStatus(InvalidWriteKey), writeKey)
+		return
+	}
+	if !misc.ContainsString(webhookSources, sourceDefName) {
+		webhook.failRequest(w, r, getStatus(InvalidWebhookSource), writeKey)
+		return
+	}
+
+	done := make(chan string)
+	req := webhookT{request: r, writer: &w, done: done, sourceType: sourceDefName, writeKey: writeKey}
+	webhook.requestQ[sourceDefName] <- &req
+
+	//Wait for batcher process to be done
+	errorMessage := <-done
+	atomic.AddUint64(&webhook.gwHandle.ackCount, 1)
+	webhook.gwHandle.trackRequestMetrics(errorMessage)
+	if errorMessage != "" {
+		logger.Debug(errorMessage)
+		http.Error(w, errorMessage, 400)
+	} else {
+		logger.Debug(getStatus(Ok))
+		w.Write([]byte(getStatus(Ok)))
+	}
+}
+
+func (webhook *webhookHandleT) batcher(sourceDef string) {
+	var reqBuffer = make([]*webhookT, 0)
+	timeout := time.After(webhookBatchTimeout)
+	for {
+		select {
+		case req := <-webhook.requestQ[sourceDef]:
+			//Append to request buffer
+			reqBuffer = append(reqBuffer, req)
+			if len(reqBuffer) == maxWebhookBatchSize {
+				breq := batchWebhookT{batchRequest: reqBuffer, sourceType: sourceDef}
+				webhook.batchRequestQ <- &breq
+				reqBuffer = nil
+				reqBuffer = make([]*webhookT, 0)
+			}
+		case <-timeout:
+			timeout = time.After(webhookBatchTimeout)
+			if len(reqBuffer) > 0 {
+				breq := batchWebhookT{batchRequest: reqBuffer, sourceType: sourceDef}
+				webhook.batchRequestQ <- &breq
+				reqBuffer = nil
+				reqBuffer = make([]*webhookT, 0)
+			}
+		}
+	}
+}
+
+func (webhook *webhookHandleT) batchTransform(process int) {
+	for breq := range webhook.batchRequestQ {
+		payloadArr := [][]byte{}
+		webRequests := []*webhookT{}
+		for _, req := range breq.batchRequest {
+			body, err := ioutil.ReadAll(req.request.Body)
+			req.request.Body.Close()
+
+			if err != nil {
+				req.done <- getStatus(RequestBodyReadFailed)
+				continue
+			}
+
+			payloadArr = append(payloadArr, body)
+			webRequests = append(webRequests, req)
+		}
+
+		if len(payloadArr) == 0 {
+			continue
+		}
+
+		// stats
+		webhook.perfStats.Start()
```

Comment:
Multiple goroutines will be trying to write to same stat.

typedef batchTransformer struct {
  webhook * webhookT
  stats *statT
}

Instead of `func (webhook *webhookHandleT) batchTransform(process int) {`
Should we have something like `func (batchTransformer *batchTransfomer) batchTransform(process int) {`